DaS_Programs/Singly linked list/Singly linked list/main.c

```diff
@@ -356,14 +362,32 @@ struct node *delete_before(struct node *start)
         
         preptr=start;
         ptr=start->next;
-        
-        while (ptr->next->data!=val)
+        if(preptr->data==val)
+		{
+			printf("\nThis is the first node. Nothing to delete\n");
+			return start;
+		}
+		if(ptr->data==val)
+		{
+			preptr=start;
+			start=start->next;
+			free(preptr);
+			return start;
+        }
+        while (ptr->next!=NULL&&ptr->next->data!=val)
         {
             preptr=ptr;
             ptr=ptr->next;
         }
-        preptr->next=ptr->next;
-        free(ptr);
+	if(ptr2->next==NULL)
+		{
+			printf("\nValue not found, Can't delete\n");
+		}
+	else
+	{
+        	preptr->next=ptr->next;
+        	free(ptr);
```

Comment:
Changes: free(ptr) comes out of the if else case.

Reply:
what is the reason for that?

Reply:
Okay my bad, I thought it was a part of the while loop as well.

Reply:
Btw boy, this case of delete before doesnt exist ha, I just wrote the code to experiment, delete after and delete node takes care of delete before situation so, delete before isnt needed.